### PR TITLE
[RLlib] Issue 39453: PettingZoo wrappers should use correct multi-agent dict action- and observation spaces.

### LINKED ISSUE #39453: [RLLib] ParallelPettingZooEnv wrapper fails rllib.utils.check_env()
### What happened + What you expected to happen

I would expect that using RLLib's own [`ParallelPettingZooEnv`](https://github.com/ray-project/ray/blob/212cd04fc68c75dc5c48c46643a9ab6cc55fd842/rllib/env/wrappers/pettingzoo_env.py) wrapper around a PettingZoo `ParallelEnv` would pass RLLib's own [`check_env()`](https://github.com/ray-project/ray/blob/212cd04fc68c75dc5c48c46643a9ab6cc55fd842/rllib/utils/pre_checks/env.py#L31) test. However, this is not the case. See the simple reproduction script below. The error output is copied below.

<details>
<summary>Error output</summary>

```
2023-09-08 08:35:40,708	WARNING multi_agent_env.py:274 -- observation_space_sample() of <ParallelPettingZooEnv instance> has not been implemented. You can either implement it yourself or bring the observation space into the preferred format of a mapping from agent ids to their individual observation spaces. 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[/usr/local/lib/python3.10/dist-packages/ray/rllib/utils/pre_checks/env.py](https://localhost:8080/#) in check_env(env, config)
     80         if isinstance(env, MultiAgentEnv):
---> 81             check_multiagent_environments(env)
     82         elif isinstance(env, VectorEnv):

3 frames
ValueError: The element returned by reset() has agent_ids that are not the names of the agents in the env. 
Agent_ids in this MultiAgentDict: ['player_0', 'player_1']
Agent_ids in this env:[]. You likely need to add the private attribute `_agent_ids` to your env, which is a set containing the ids of agents supported by your env.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
[/usr/local/lib/python3.10/dist-packages/ray/rllib/utils/pre_checks/env.py](https://localhost:8080/#) in check_env(env, config)
     94     except Exception:
     95         actual_error = traceback.format_exc()
---> 96         raise ValueError(
     97             f"{actual_error}\n"
     98             "The above error has been found in your environment! "

ValueError: Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/ray/rllib/utils/pre_checks/env.py", line 81, in check_env
    check_multiagent_environments(env)
  File "/usr/local/lib/python3.10/dist-packages/ray/rllib/utils/pre_checks/env.py", line 320, in check_multiagent_environments
    _check_if_element_multi_agent_dict(env, reset_obs, "reset()")
  File "/usr/local/lib/python3.10/dist-packages/ray/rllib/utils/pre_checks/env.py", line 785, in _check_if_element_multi_agent_dict
    raise ValueError(error)
ValueError: The element returned by reset() has agent_ids that are not the names of the agents in the env. 
Agent_ids in this MultiAgentDict: ['player_0', 'player_1']
Agent_ids in this env:[]. You likely need to add the private attribute `_agent_ids` to your env, which is a set containing the ids of agents supported by your env.

The above error has been found in your environment! We've added a module for checking your custom environments. It may cause your experiment to fail if your environment is not set up correctly. You can disable this behavior via calling `config.environment(disable_env_checking=True)`. You can run the environment checking module standalone by calling ray.rllib.utils.check_env([your env]).
```

</details>

## Root Issue

The root issue for this error is that RLLib's `ParallelPettingZooEnv` wrapper does not adopt the so-called "preferred format" in RLLib's own [`MultiAgentEnv`](https://github.com/ray-project/ray/blob/master/rllib/env/multi_agent_env.py) API

https://github.com/ray-project/ray/blob/212cd04fc68c75dc5c48c46643a9ab6cc55fd842/rllib/env/wrappers/pettingzoo_env.py#L89-L114

## Recommendations

1. Update the `ParallelPettingZooEnv` wrapper to use the preferred format.

2. Add a unit test in [test_pettingzoo_env.py](https://github.com/ray-project/ray/blob/master/rllib/tests/test_pettingzoo_env.py) to ensure that a wrapped PettingZoo passes RLLib's own `check_env()` function.

## Why this is important

As a creator of various multi-agent RL environments following the PettingZoo API, I would love to be able to use RLLib's `check_env()` function as a unit test / sanity check for my own environment to ensure that users can train RLLib algorithms on my environment. Ideally, I'd like the following to work:

Implementing a new multi-agent RL environment:
```python
from pettingzoo.utils.env import ParallelEnv

class MyEnv(ParallelEnv):
    ...
```

Unit test
```python
import unittest

from ray.rllib.env.wrappers.pettingzoo_env import ParallelPettingZooEnv
import ray.rllib.utils

class TestMyEnv(unittest.TestCase):
    def test_rllib_check_env(self):
        env = MyEnv()
        rllib_env = ParallelPettingZooEnv(env)
        ray.rllib.utils.check_env(rllib_env)
```

### Versions / Dependencies

ray[rllib]==2.6.3
pettingzoo==1.22.3  (this is the latest version of pettingzoo that ray[rllib]==2.6.3 is compatible with)

### Reproduction script

Run the following in any Jupyter notebook or Google Colab notebook:

```python
!pip install "ray[rllib]==2.6.3" pettingzoo[classic]

from pettingzoo.classic import rps_v2
from ray.rllib.env.wrappers.pettingzoo_env import ParallelPettingZooEnv
import ray.rllib.utils

env = rps_v2.parallel_env(render_mode="human")
rllib_env = ParallelPettingZooEnv(env)
ray.rllib.utils.check_env(rllib_env)
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by sven1977**
See issue: #39453 

Solution:
Change both PettingZoo adapters of RLlib to abide to RLlib's multi-agent action- and observation space convention (gym.spaces.Dict mapping agent Ids to individual spaces).

Closes: #39453 and #32889

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by chrisyeh96** [comment]
Thanks @sven1977 for responding so quickly to #39453!

I have a suggestion: can we make RLLib's `PettingZooEnv` and `ParallelPettingZooEnv` wrappers support PettingZoo environments where agents have different `action_spaces` and `observation_spaces`? This would simplify the RLLib wrapper code quite significantly.

For example, here's what the `__init__()` method for `ParallelPettingZooEnv` would look like:

```python
class ParallelPettingZooEnv(MultiAgentEnv):
    def __init__(self, env):
        super().__init__()
        self.par_env = env
        self.par_env.reset()
        self._agent_ids = set(self.par_env.agents)

        self.observation_space = gym.spaces.Dict(self.par_env.observation_spaces)
        self.action_space = gym.spaces.Dict(self.par_env.action_spaces)
```

**Comment by sven1977** [comment]
Hey @chrisyeh96 , thanks for the hint. Changed this to be even more flexible now for both parallel and non-parallel env wrappers.

File: rllib/env/wrappers/pettingzoo_env.py [hunk 1]
```diff
@@ -82,58 +84,22 @@ def __init__(self, env):
...
-                    "`supersuit.aec_wrappers.pad_action_space(env)`"
-                )
-
-        # Convert from gym to gymnasium, if necessary.
-        self.observation_space = convert_old_gym_space_to_gymnasium_space(
-            first_obs_space
-        )
-        self.action_space = convert_old_gym_space_to_gymnasium_space(first_action_space)
-
         self._agent_ids = self.env.agents
```
Comment (chrisyeh96) [inline rllib/env/wrappers/pettingzoo_env.py:87]: `self._agent_ids` should be a set, so this line should be changed to:

```python
self._agent_ids = set(self.env.agents)
```

File: rllib/env/wrappers/pettingzoo_env.py [hunk 2]
```diff
@@ -188,37 +154,10 @@ def __init__(self, env):
         super().__init__()
         self.par_env = env
         self.par_env.reset()
+        self._agent_ids = self.par_env.agents
```
Comment (chrisyeh96) [inline rllib/env/wrappers/pettingzoo_env.py:157]: `self._agent_ids` should be a set, so this line should be changed to:

```python
self._agent_ids = set(self.par_env.agents)
```

**Review by chrisyeh96 — CHANGES_REQUESTED** [review]

**Comment by chrisyeh96** [comment]
Thanks @sven1977 for the changes! I've suggested 2 minor changes in my code review.

There's also one other place that should be updated. The docstring for `PettingZooEnv` currently still says that the wrapper only supports environments with homogeneous action / observation spaces. With this pull request, this comment should be removed. See below:

https://github.com/ray-project/ray/blob/5a4389be43a3d432d4c91cfbeb14a8cd62a26453/rllib/env/wrappers/pettingzoo_env.py#L23-L27

Thanks for your work on this!

File: rllib/env/wrappers/pettingzoo_env.py (hunk 2 above)
Comment (sven1977) [reply rllib/env/wrappers/pettingzoo_env.py]: done

File: rllib/env/wrappers/pettingzoo_env.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/wrappers/pettingzoo_env.py]: done

**Comment by chrisyeh96** [comment]
Thanks @sven1977! Looking forward to seeing this change merged. Do you know if it will be included in the next release of RLLib?

**Comment by chrisyeh96** [comment]
Understood. Any idea on what the timeline for Ray 2.8 will be? For me, this issue is quite important.

I have a project called [SustainGym](https://github.com/chrisyeh96/sustaingym) where I've made 5 RL environments related to energy / sustainability applications. 3 of the 5 environments are multi-agent RL environments following the PettingZoo API.

Currently, I include RLLib as a dependency of SustainGym, because I have special versions of my environments to fit the RLLib `MultiAgentEnv` API. In theory, SustainGym should not depend on RLLib, because a RL environment package should be agnostic to the RL algorithm. Once this pull request gets merged, I can remove the RLLib dependency and instead direct SustainGym users to simply use the RLLib wrapper around the SustainGym PettingZoo env if they want to use RLLib algorithms.

**Comment by sven1977** [comment]
2.8 will be released a few weeks after the Ray Summit (which is next week). So roughly beginning to mid October.

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**